### DISCUSSION TITLE
"Overload 2 is the closest match" when it is not

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#7292] [General]
From [this StackOverflow question](https://stackoverflow.com/q/78017822):

```py
results = {"items": [{"item_id": "Foo"}, {"item_id": "Bar"}]}
reveal_type(results)          # dict[str, list[dict[str, str]]]

results.update({"q": "bar"})  # error: No overloads for "update" match the provided arguments
                              # (Pylance) Overload 2 is the closest match
```

![](https://github.com/microsoft/pyright/assets/122007197/a689e12f-98eb-4bf5-9f01-3f77ee9c627d)

The overloads are defined as:

```py
class MutableMapping(Mapping[_KT, _VT]):
    ...
    @overload
    def update(self, __m: SupportsKeysAndGetItem[_KT, _VT], **kwargs: _VT) -> None: ...
    @overload
    def update(self, __m: Iterable[tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
    @overload
    def update(self, **kwargs: _VT) -> None: ...
```

I'd argue the closest match is overload 1 and not 2.

**Comment by erictraut** [comment]
Pyright is working as designed here. "Closest" is subjective, and pyright uses some heuristics to attempt to determine the overload that was most likely intended by the user given the specified arguments. These heuristics have been tuned iteratively over time. Any further changes are likely to be seen as regressions from the current heuristics, so I'm reluctant to make any changes here.
